iocage/lib/ioc_create.py

```diff
@@ -86,6 +86,106 @@ def create_jail(self):
+                    "message": "You cannot clone a template, "
+                               "use create -t instead."
+                },
+                    _callback=self.callback,
+                    silent=self.silent)
+            else:
+                # Yep, self.release is actually the source jail.
+                iocage.lib.ioc_common.logit({
+                    "level"  : "EXCEPTION",
+                    "message": f"Jail: {self.release} not found!"
+                },
+                    _callback=self.callback,
+                    silent=self.silent)
+        return config_fn
+
+    def _get_release_config(self, jail_uuid, path):
+        freebsd_version = f"{path}/root/bin/freebsd-version"
+
+        if self.empty:
+            return "EMPTY"
+
+        if self.release[:4].endswith("-"):
+            # 9.3-RELEASE and under don't actually have this binary.
+            return self.release
+
+        try:
+            with open(freebsd_version, "r") as r:
+                for line in r:
+                    if line.startswith("USERLAND_VERSION"):
+                        # Long lines ftw?
+                        cloned_release = line.rstrip().partition("=")[2]
+                        return cloned_release.strip('"')
+
+        except (IOError, OSError, FileNotFoundError, UnboundLocalError):
+            iocage.lib.ioc_common.logit({
+                "level"  : "EXCEPTION",
+                "message": f"RELEASE: {self.release} not found!"
+            },
+                _callback=self.callback,
+                silent=self.silent)
+
+    def _su_check_call(self, jail_uuid, path, _type):
+        if self.empty:
+            return
+
+        rec = "-r" if _type == "clone" else ""
+
+        if _type == "template":
+            dataset = f"{path}/root@{jail_uuid}"
+            err = f"Template {self.release} not found!"
+        if _type == "releases":
+            dataset = f"{path}/root@{jail_uuid}"
+            err = f"RELEASE {self.release} not found!"
+        elif _type == "clone":
+            dataset = f"{path}@{jail_uuid}"
+            err = f"Jail {jail_uuid} not found!"
+
+        try:
+            su.check_call(["zfs", "snapshot", rec, dataset], stderr=su.PIPE)
+        except su.CalledProcessError:
+            try:
+                snapshot = self.zfs.get_snapshot(dataset)
+                iocage.lib.ioc_common.logit({
+                    "level"  : "EXCEPTION",
+                    "message": f"Snapshot: {snapshot.name} exists!\n"
+                               "Please manually run zfs destroy"
+                               f" {snapshot.name} if you wish to "
+                               "destroy it."
+                },
+                    _callback=self.callback,
+                    silent=self.silent)
+
+            except libzfs.ZFSException:
+                pass
+            raise RuntimeError(err)
+
+    def _zfs_clone(src, dst, *opts):
+        su.Popen(["zfs", "clone", opts, src, dst],
```

Comment:
puppet has a really nice concept of creating "commands" like so:

```ruby
  commands zfs: "zfs"
```
which basically wraps the 80% case of just calling `exec('zfs', params)`
but… more important might be wether `*opts` is the way to go for python ;)